Dynamic pvp centers.

StrawWagen opened the pull request:
Change spawning algorithm's handling of "No predefined pvp centers" case

Now creates a new pvp center at the average point of all pvpers. 
This center is rebuilt at the new average point every update interval
Dynamic center needs a radius at least DYNAMIC_CENTER_MINDIST in size. With at least DYNAMIC_CENTER_MINSPAWNS inside of said radius.

+ automatically snap cfc_spawneditor_spawnadd spawn's yaws

StrawWagen committed:
Dynamic pvp centers.

Change spawning algorithm's handling of "No predefined pvp centers"

Now creates a new pvp center at the average point of all pvpers.
Dynamic center needs a radius at least DYNAMIC_CENTER_MINDIST in size.
With at least DYNAMIC_CENTER_MINSPAWNS inside of said radius.

StrawWagen committed:
Polish dynamic centers.

StrawWagen committed:
Add a sprinkle of randomness

+Finalized tdm_outercanals  spawns

StrawWagen commented:
![20231027144136_1](https://github.com/CFC-Servers/cfc_random_spawn/assets/64710817/8649d120-0d01-4527-93ba-9d0a77a436f5)
![20231027144147_1](https://github.com/CFC-Servers/cfc_random_spawn/assets/64710817/e330ba3c-8860-45b8-b0d4-cfc5b3b22bec)
![20231027144144_1](https://github.com/CFC-Servers/cfc_random_spawn/assets/64710817/04aa1499-b172-4f00-8c9a-f0da248b4e23)
![20231027144140_1](https://github.com/CFC-Servers/cfc_random_spawn/assets/64710817/6798cc4b-14d7-4a09-abae-e5cc3bd7a687)

New tdm_outercanals spawns that come with this PR

StrawWagen committed:
Improve comments

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -14,21 +14,20 @@ local CENTER_UPDATE_INTERVAL = customSpawnConfigForMap.centerUpdateInterval or C
+local DYNAMIC_CENTER_SPAWNCOUNTMATCHPVPERS = true -- pvp center gets bigger when more people are pvping
+local DYNAMIC_CENTER_IMPERFECT = true -- throw a bit of randomness in, makes pvp less stiff.
+
+local function defaultPvpCenter()
+    if pvpCenters[1] then return pvpCenters[1] end
```
```suggestion
local function defaultPvpCenter()
    return pvpCenters[1]
```
could simplify to this

StrawWagen replied:
👍

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -105,22 +103,43 @@ local function getLivingPlayers()
+
+    if not nearestSpawn then return spawns[1] end
+
+    return nearestSpawn
```
you could probably  delete this because you should always have a nearestSpawn set starting with math.huge
```suggestion
    return nearestSpawn
```

StrawWagen replied:
👍

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -197,13 +216,74 @@ local function getPlayerPopularityFromPoint( point, plys, radiusSqr )
+local function getDynamicPvpCenter( measurablePlayers )
+    local playersAveragePos = Vector( 0, 0, 0 )
+    for _, ply in ipairs( measurablePlayers ) do
+        playersAveragePos = playersAveragePos + ply:GetPos()
```
theres already a getPlyAvg function used other places in the code. May be better to use that.

The getPlyAvg function itself could also probably be improved. It could not take count manually and use the `#` operator.

StrawWagen replied:
getplyavg used count instead of #, to enable some specific behaviour iirc, something about picking the best pvpcenter when there's multiple

StrawWagen replied:
That functionality wasn't as important as i thought. Applied this suggestion!

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -217,20 +297,28 @@ local function getPopularCenter( plys )
 
     local freeSpawns = findFreeSpawnPoints( customSpawnsForMap, allLivingPlys )
-    local nearestSpawns = getNearestSpawns( getPlyAvg( measurablePlayers, mostPopularCenter.centerPos ), freeSpawns )
+    local nearestSpawns = getNearestSpawns( getPlyAvg( measurablePlayers, CFCRandomSpawn.mostPopularCenter.centerPos ), freeSpawns )
```
This also isnt in the scope of this PR but it looks like every spawn. We are calculating if every spawn is free, then getting the nearest spawns, then sorting them. 

I think most spawns will usually be free. So depending on how many spawns exist It might be more efficient to instead get the nearest spawns without calculating if they are free, pick a random one, check if its free, then repeat until you get a free spawn.

plally replied:
Then we could cache the nearest spawns table for a few seconds and avoid recalculating it every spawn.

StrawWagen replied:
good changes but they're out of scope for now.
but i'll make findFreeSpawnPoints into findFreeSpawnPoint and have it run after getNearestSpawns

StrawWagen replied:
Nearest spawns table is also cached now.

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -217,20 +297,28 @@ local function getPopularCenter( plys )
-        CENTER_CUTOFF = mostPopularCenter.overrideCutoff or DEFAULT_CENTER_CUTOFF
-        CENTER_CUTOFF_SQR = mostPopularCenter.overrideCutoffSqr or DEFAULT_CENTER_CUTOFF_SQR
+    if CENTER_UPDATE_ON_RESPAWN or not CFCRandomSpawn.mostPopularCenter then
+        updatePopularCenter( measurablePlayers )
```
could we just get rid of the option to update center on respawn? is there every a time we will want to do that?

StrawWagen replied:
👍

StrawWagen committed:
Apply suggestions from code review

Co-authored-by: Pierce Lally <pierce@vulpes.dev>

StrawWagen committed:
Apply more suggestions from code review + Extra

Decreased "CLOSENESS_LIMIT" in config, was 400, now is 100

Removed "CENTER_UPDATE_ON_RESPAWN" functionality
 Changed "getNearestSpawns" to "spawnsInsidePvpCenterCached"

Simplify getPlyAvg

getOptimalSpawnPos no longer calls getPlyAvg, spawnsInsidePvpCenterCached doesn't need it.

Change "getPopularCenter" to "getMostPopulatedCenter" as the variable "mostPopularCenter" isn't only set by that function

No longer find all free spawnpoints, just return the first free one.

StrawWagen committed:
Fix style

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -105,80 +101,107 @@ local function getLivingPlayers()
-                break
-            end
-        end
+    for _, _ in pairs( spawns ) do
```
```suggestion
    for _=1, #spawns do
```

plally replied:
pairs might break when youre removing  stuff while iterating

StrawWagen replied:
👍

plally commented on lua/cfc_random_spawn/module/sv_player_spawning.lua:
```diff
@@ -105,80 +101,107 @@ local function getLivingPlayers()
+        else
+            -- remove this spawn!
+            -- TODO, operate on the table randomly instead of this lil bit of jankiness.
+            table.remove( spawns, randomIndex )
```
```suggestion
            spawns[randomIndex] = spawns[#spawns]
            spawns[#spawns] = nil
```

StrawWagen replied:
👍

StrawWagen committed:
Apply suggestions from code review

Co-authored-by: Pierce Lally <pierce@vulpes.dev>

StrawWagen committed:
Apply more suggestions.

Tested, works!
